elixir/apps/domain/lib/domain/auth/adapters/okta/api_client.ex

```diff
@@ -190,4 +202,55 @@ defmodule Domain.Auth.Adapters.Okta.APIClient do
   end
 
   defp parse_link_header(nil), do: nil
+
+  defp parse_headers_for_errors(headers) do
+    headers
+    |> Enum.find({}, fn {key, _val} -> key == "www-authenticate" end)
+    |> parse_error_header()
+  end
+
+  def parse_error_header({"www-authenticate", errors}) do
+    String.split(errors, ",")
+    |> Enum.map(&String.trim/1)
+    |> Enum.filter(&String.starts_with?(&1, "error"))
+    |> Enum.map(&String.replace(&1, "\"", ""))
+    |> Enum.map(&String.split(&1, "="))
+    |> Enum.into(%{}, fn [a, b] -> {a, b} end)
+  end
+
+  def parse_error_header(_) do
+    Logger.info("No www-authenticate header present")
+    %{"error" => "unknown", "error_message" => "no www-authenticate header present"}
+  end
+
+  defp fetch_latest_access_token(provider) do
+    access_token = provider.adapter_state["access_token"]
+
+    if is_access_token_active?(access_token) do
+      access_token
+    else
+      # Fetch provider from DB and return latest access token
+      provider = Domain.Auth.fetch_active_provider_by_id(provider.id)
+      provider.adapter_state["access_token"]
+    end
+  end
+
+  defp is_access_token_active?(token) do
+    # Using try/rescue because JOSE.JWT.peek could fail if access token is not JWT
+    try do
+      jwt = JOSE.JWT.peek(token)
+      current_time = DateTime.utc_now()
+      {:ok, timestamp_time} = DateTime.from_unix(jwt.fields["exp"])
+
+      case DateTime.compare(current_time, timestamp_time) do
+        :lt -> true
+        :eq -> false
+        :gt -> false
+      end
+    rescue
+      _error ->
+        Logger.warning("Error while checking JWT access token")
+        false
+    end
+  end
```

Comment:
Probably not the end of the world. I know this is an anti-pattern in Elixir though.

If we stick with this try/rescue approach I would just keep it limited to only the call that can raise to minimize the chance of sweeping up other exceptions here. And log the `inspect(exception)` in the Logger.error call.

Reply:
Yeah, I know this is not a very popular thing in Elixir and I hesitated to use it here.  I just wasn't sure how likely it was that we might end up with the provider state having an access token that was either an empty string or `nil` due to not being there at all and didn't want this to crash everything.

After having stepped away from it for a little bit, I think I've got an idea of how I can do this without needing a try/rescue, so I'll get it updated.

Reply:
I've updated this to remove the try/rescue.  Let me know what you think.